### PR TITLE
sensor: mcp9808: adding ability to change the resolution of the sensor through kconfig

### PR DESCRIPTION
**Pull request by sdaglish**
The current MCP9808 driver does not set the resolution for the sensor, instead using the default resolution of 0.0625C.

In some cases, it is better to use a lower resolution (for example to reduce the time required to acquire a new sample).

This PR adds the ability to change the resolution of the temperature measurements at compile time.

As the resolution is set during the MCP9808 init function, the i2c write functions have been moved from mcp9808_trigger.c to mcp9808.c.

Since the resolution register is only 8 bits, an 8bit write function has been created, and the original 16bit write function renamed to show the difference.

### CODE REVIEW

File: drivers/sensor/mcp9808/mcp9808.h [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/*
+*
```
Comment (MaureenHelm) [inline drivers/sensor/mcp9808/mcp9808.h]: this doesn't look right...

File: drivers/sensor/mcp9808/mcp9808.h [hunk 2]
```diff
@@ -16,11 +16,11 @@
...
-#define MCP9808_REG_CONFIG		0x01
-#define MCP9808_REG_UPPER_LIMIT		0x02
-#define MCP9808_REG_LOWER_LIMIT		0x03
-#define MCP9808_REG_CRITICAL		0x04
-#define MCP9808_REG_TEMP_AMB		0x05
+#define MCP9808_REG_CONFIG              0x01
+#define MCP9808_REG_UPPER_LIMIT         0x02
+#define MCP9808_REG_LOWER_LIMIT         0x03
+#define MCP9808_REG_CRITICAL            0x04
+#define MCP9808_REG_TEMP_AMB            0x05
```
Comment (MaureenHelm) [inline drivers/sensor/mcp9808/mcp9808.h:23]: The whitespace was fine, why do you want to change it?

File: samples/sensor/mcp9808/boards/nucleo_l053r8.overlay [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+ */
+
+&i2c1 {
+	mcp9808@18 {
+		compatible = "microchip,mcp9808";
+		reg = <0x18>;
+		int-gpios = <&gpioa 8 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+		label = "MCP9808";
+	};
+};
```
Comment (MaureenHelm) [inline samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:14]: Please add this board to the `sample.yaml`

File: drivers/sensor/mcp9808/Kconfig [hunk 4]
```diff
@@ -30,6 +30,24 @@ config MCP9808_TRIGGER_OWN_THREAD
...
+config MCP9808_MEAS_RES_0_25C
+	bool "0.25C temperature resolution"
+
+config MCP9808_MEAS_RES_0_125C
+	bool "0.125C temperature resolution"
+
+config MCP9808_MEAS_RES_0_0625C
+	bool "0.0625C temperature resolution"
+
+endchoice
```
Comment (MaureenHelm) [inline drivers/sensor/mcp9808/Kconfig:49]: We have been migrating toward defining these as dts properties instead of Kconfigs, as this allows different selections per instance.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for the PR!

Please move the resolution selection from a Kconfig choice to a device tree property.

I don't think the whitespace commit is necessary.

File: drivers/sensor/mcp9808/Kconfig (hunk 4 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/Kconfig:49]: Is there an example of this being used in one of the sample projects? I'm struggling to find an example of this implementation to work from.

File: drivers/sensor/mcp9808/mcp9808.h (hunk 1 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.h]: This was fixed in the latest push.

File: drivers/sensor/mcp9808/Kconfig (hunk 4 above)
Comment (MaureenHelm) [reply drivers/sensor/mcp9808/Kconfig:49]: See #30317

**Review by sdaglish — COMMENTED** [review]
Thanks for the example commit. I have now moved the resolution selection into the dts instead of the kconfig.

File: samples/sensor/mcp9808/boards/nucleo_l053r8.overlay (hunk 3 above)
Comment (sdaglish) [reply samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:15]: The board has now been added to `sample.yaml`

File: dts/bindings/sensor/microchip,mcp9808.yaml [hunk 5]
```diff
@@ -18,3 +18,14 @@ properties:
         sensor, and is open-drain.  A pull-up may be appropriate.  The
         property value should ensure the flags properly describe the
         signal that is presented to the driver.
+
+    resolution:
+      type: int
+      required: false
+      default: 3
```
Comment (MaureenHelm) [inline dts/bindings/sensor/microchip,mcp9808.yaml:25]: Is there a reason you picked this default value? Is it the reset value in the sensor register? Please document where the default came from in the description.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> I have now moved the resolution selection into the dts instead of the kconfig.

Thank you. This part looks good.

File: dts/bindings/sensor/microchip,mcp9808.yaml (hunk 5 above)
Comment (sdaglish) [reply dts/bindings/sensor/microchip,mcp9808.yaml:25]: This is the default, power-on, value. This information has now been documented.

**Review by MaureenHelm — APPROVED** [review]
Looks good, thank you.

Please review and try to fix the buildkite failures.

File: drivers/sensor/mcp9808/mcp9808.c [hunk 6]
```diff
@@ -82,6 +115,12 @@ int mcp9808_init(const struct device *dev)
 		return -EINVAL;
 	}
 
+	rc = mcp9808_set_temperature_resolution(dev, cfg->resolution);
+	if (!rc) {
+		LOG_DBG("Could not set the resolution of mcp9808 module");
```
Comment (avisconti) [inline drivers/sensor/mcp9808/mcp9808.c:120]: Should not be LOG_ERR()?

File: drivers/sensor/mcp9808/mcp9808.c [hunk 7]
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
...
+	uint8_t buf[3] = {
+		reg,
+		val,
+	};
+
+	return i2c_write(data->i2c_master, buf, sizeof(buf), cfg->i2c_addr);
+}
+
+static int mcp9808_set_temperature_resolution(const struct device *dev,
+					      uint8_t resolution)
```
Comment (avisconti) [inline drivers/sensor/mcp9808/mcp9808.c:66]: I'm wondering if this is the correct indentation. I think 'uint8_t' should be aligned under 'const'. But I also see that this would break the other indentations in the driver. I don't know what to suggest, if leaving as it is in accordance with the rest of the driver, or to fix it.

File: drivers/sensor/mcp9808/mcp9808.c [hunk 8]
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
...
+		val & 0xFF,
+	};
+
+	return i2c_write(data->i2c_master, buf, sizeof(buf), cfg->i2c_addr);
+}
+
+int mcp9808_reg_write_8bit(const struct device *dev, uint8_t reg,
+			   uint8_t val)
+{
+	const struct mcp9808_data *data = dev->data;
```
Comment (avisconti) [inline drivers/sensor/mcp9808/mcp9808.c:55]: Is data struct really 'const'?

File: drivers/sensor/mcp9808/mcp9808.c [hunk 9]
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
 	return rc;
 }
 
+int mcp9808_reg_write_16bit(const struct device *dev, uint8_t reg,
+			    uint16_t val)
+{
+	const struct mcp9808_data *data = dev->data;
```
Comment (avisconti) [inline drivers/sensor/mcp9808/mcp9808.c:41]: Is data struct really 'const'?

**Review by avisconti — CHANGES_REQUESTED** [review]
LGTM, just minor things

File: drivers/sensor/mcp9808/mcp9808.c (hunk 6 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:120]: Agreed and amended.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 8 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:55]: I believe that in this instance it should be `const`. `i2c_write` requires `const struct device *dev` as first parameter.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 9 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:41]: I believe that in this instance it should be `const`. `i2c_write` requires `const struct device *dev` as first parameter.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 7 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:66]: Happy with either convention. Small personal preference to align `uint8_t` with `const`. That is how it's showing on my device, but I need to change the indentation correctly. I'll update this and re-push.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 9 above)
Comment (gmarull) [reply drivers/sensor/mcp9808/mcp9808.c:41]: driver `data` is never `const`, devices are.

File: drivers/sensor/mcp9808/mcp9808.c [hunk 10]
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
...
+	const struct mcp9808_config *cfg = dev->config;
+	uint8_t buf[3] = {
+		reg,
+		val >> 8,       /* big-endian register storage */
```
Comment (gmarull) [inline drivers/sensor/mcp9808/mcp9808.c:45]: use facilities in `byteorder.h`

File: drivers/sensor/mcp9808/mcp9808.c [hunk 11]
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
...
+
+	return i2c_write(data->i2c_master, buf, sizeof(buf), cfg->i2c_addr);
+}
+
+int mcp9808_reg_write_8bit(const struct device *dev, uint8_t reg,
+			   uint8_t val)
+{
+	const struct mcp9808_data *data = dev->data;
+	const struct mcp9808_config *cfg = dev->config;
+	uint8_t buf[3] = {
```
Comment (gmarull) [inline drivers/sensor/mcp9808/mcp9808.c:57]: shouldn't it be `2`?

File: drivers/sensor/mcp9808/mcp9808.c [hunk 12]
```diff
@@ -78,10 +111,16 @@ int mcp9808_init(const struct device *dev)
 
 	data->i2c_master = device_get_binding(cfg->i2c_bus);
 	if (!data->i2c_master) {
-		LOG_DBG("mcp9808: i2c master not found: %s", cfg->i2c_bus);
+		LOG_ERR("mcp9808: i2c master not found: %s", cfg->i2c_bus);
 		return -EINVAL;
 	}
 
+	rc = mcp9808_set_temperature_resolution(dev, cfg->resolution);
+	if (!rc) {
```
Comment (gmarull) [inline drivers/sensor/mcp9808/mcp9808.c:119]: This eval will be true for rc == 0, which is not error. Have you tested these changes?

File: drivers/sensor/mcp9808/mcp9808.c [hunk 13]
```diff
@@ -97,9 +136,10 @@ static const struct mcp9808_config mcp9808_cfg = {
...
 	.alert_controller = DT_INST_GPIO_LABEL(0, int_gpios),
+	.resolution = DT_INST_PROP(0, resolution),
 #endif /* CONFIG_MCP9808_TRIGGER */
 };
 
 DEVICE_DT_INST_DEFINE(0, mcp9808_init, device_pm_control_nop,
-		    &mcp9808_data, &mcp9808_cfg, POST_KERNEL,
-		    CONFIG_SENSOR_INIT_PRIORITY, &mcp9808_api_funcs);
+		      &mcp9808_data, &mcp9808_cfg, POST_KERNEL,
+		      CONFIG_SENSOR_INIT_PRIORITY, &mcp9808_api_funcs);
```
Comment (gmarull) [inline drivers/sensor/mcp9808/mcp9808.c:145]: is this change needed to align?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/mcp9808/mcp9808.c (hunk 11 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:57]: That is correct and amended.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 10 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:45]: `const` removed from data.

I'm struggling to figure out why using facilities from `byteorder.h` will make this small section any better. It's already clear what the section is doing.

I've read through 'byteroder.h' and even after changing the order, I would still need to manually place the different bytes within the array.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 12 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:119]: This was changed at request: https://github.com/zephyrproject-rtos/zephyr/pull/31758#discussion_r568488344

Please confirm if this is to be set as an error or not. If not, I'll put it back to log debug.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 9 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c]: Updated

File: drivers/sensor/mcp9808/mcp9808.c (hunk 12 above)
Comment (gmarull) [reply drivers/sensor/mcp9808/mcp9808.c:119]: the problem is not `LOG_ERR`, this branch will be executed if rc == 0, i.e. operation succeeds.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 10 above)
Comment (gmarull) [reply drivers/sensor/mcp9808/mcp9808.c:45]: you can use `sys_put_be16` to insert the value

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/mcp9808/mcp9808.c (hunk 12 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:119]: After testing with logging enabled, I fully see what you mean. This will be updated.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 10 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:45]: Sorry, but I'm still not seeing this. I can use the following to include `sys_put_be16`, but I still don't see it as being any better:

`uint8_t buff[3] = {reg};
sys_put_be16(val, &buf[1]);`

It compiles and runs, but I'm not seeing the benefit.

Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:45]: Or

`uint8_t buff[3];
buff[0] = reg;
sys_put_be16(val, &buf[1]);`

Comment (gmarull) [reply drivers/sensor/mcp9808/mcp9808.c:45]: I think it is better to use system facilities when available, they'll usually be tested and in some cases optimized.

Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c]: Updated to use `sys_put_be16`

File: samples/sensor/mcp9808/sample.yaml [hunk 14]
```diff
@@ -6,3 +6,4 @@ tests:
     tags: sensors
     depends_on: i2c
     filter: dt_compat_enabled("microchip,mcp9808")
+    platform_allow: nucleo_l053r8
```
Comment (gmarull) [inline samples/sensor/mcp9808/sample.yaml:9]: not sure if this is strictly needed

File: samples/sensor/mcp9808/boards/nucleo_l053r8.overlay [hunk 15]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2019, Linaro Ltd.
```
Comment (gmarull) [inline samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:2]: Copyright should likely be adjusted (doesn't seem to be copied?)

**Review by gmarull — APPROVED** [review]
Added a couple of minor comments, but the rest LGTM. Thanks!

File: samples/sensor/mcp9808/boards/nucleo_l053r8.overlay (hunk 15 above)
Comment (sdaglish) [reply samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:2]: Fixed.

File: drivers/sensor/mcp9808/mcp9808.c (hunk 13 above)
Comment (sdaglish) [reply drivers/sensor/mcp9808/mcp9808.c:145]: This is the output from the default uncrustify. I presume this change is correct.

File: samples/sensor/mcp9808/sample.yaml (hunk 14 above)
Comment (sdaglish) [reply samples/sensor/mcp9808/sample.yaml:9]: Not necessary, but I think good to add.

Addition was requested here: https://github.com/zephyrproject-rtos/zephyr/pull/31758#discussion_r566399165

**Review by avisconti — APPROVED** [review]
LGTM now
